include/storage/io.hpp

```diff
@@ -20,11 +24,134 @@ namespace storage
 namespace io
 {
 
+class File
+{
+  private:
+    std::string filename;
+    boost::filesystem::ifstream input_stream;
+
+  public:
+    File(const std::string &filename, const bool check_fingerprint = false)
+        : File(boost::filesystem::path(filename), check_fingerprint)
+    {
+    }
+
+    File(const boost::filesystem::path &filename_, const bool check_fingerprint = false)
+    {
+        filename = filename_.string();
+        input_stream.open(filename_, std::ios::binary);
+        if (!input_stream)
+            throw util::exception("Error opening " + filename + ":" + std::strerror(errno));
+
+        if (check_fingerprint && !readAndCheckFingerprint())
+        {
+            throw util::exception("Fingerprint mismatch in " + filename);
+        }
+    }
+
+    /* Read count objects of type T into pointer dest */
+    template <typename T> void readInto(T *dest, const std::size_t count)
```

Comment:
Ref is not possible due to reading multiple entries, but can we use a iterator range and check for overflowing the destination?